packages/explat-client/src/internal/local-storage.ts

```diff
@@ -20,18 +21,22 @@ export const polyfilledLocalStorage = {
 	get length(): number {
 		return Object.keys( this._data ).length;
 	},
-	key: function ( index: number ): string | undefined {
-		return Object.keys( this._data )[ index ];
+	key: function ( index: number ): string | null {
+		const keys = Object.keys( this._data );
+		return index in keys ? keys[ index ] : null;
 	},
 };
 
 /**
  * A polyfilled LocalStorage.
  * The polyfill is required at least for testing.
  */
-const localStorage =
-	typeof window !== 'undefined' && window.localStorage
-		? window.localStorage
-		: polyfilledLocalStorage;
+let localStorage = polyfilledLocalStorage;
+try {
+	localStorage =
+		typeof window !== 'undefined' && window.localStorage
+			? window.localStorage
+			: polyfilledLocalStorage;
```

Comment:
Can be now written a bit better like
```js
let localStorage = polyfilledLocalStorage;
try {
  if ( typeof window !== 'undefined' ) {
    localStorage = window.localStorage;
  }
} catch {
  // leave as polyfilled
}

Reply:
Couldn't `window.localStorage` be undefined causing localStorage to be undefined?
Maybe changing the inner block to:
```
if (window.localStorage) {
    localStorage = window.localStorage;
}
```

Reply:
Oh yes, I don't know if there's really a browser environment that has `window` but doesn't define `window.localStorage`, but let's check it. My point was more about doing the `polyfilledLocalStorage` assignment just once.

There's another very common situation where the polyfill should be detected and used: it's when the `localStorage` object is accessible, but all its methods (like `getItem`) throw a `SecurityError`. Chrome does that in Incognito mode where third-party cookies and local storage are disabled.

We discussed this recently, in https://github.com/Automattic/wp-calypso/pull/73525#issuecomment-1437592820, on another storage-access-fixing PR, with @sirbrillig, @noahtallen and @danielbachhuber, but never got back to fixing this case.

Our codebase is full of unguarded `window.localStorage.getItem` references, but they are particularly dangerous in logged-out mode, like during signup. For logged-in users, it seems their browsers tend to be more "normal".